design/general_refactorings.md

```diff
@@ -0,0 +1,381 @@
+      ```
+
+    * `RuntimeError` should be replaced with `InvalidURLException`
+
+      ```python
+      if url.scheme != 'https':
+          raise RuntimeError(f"Error: undefined behavior. Reason: The Conjur URL format provided "
+                 f"'{self.conjurrc_data.appliance_url}' is not supported.")
+      ```
+
+    * `RuntimeError` should be replaced with `ConfirmationException` 
+
+      ```python
+      trust_certificate = input("Trust this certificate? (Default=no): ").strip()
+      if trust_certificate.lower() != 'yes':
+          raise RuntimeError("You decided not to trust the certificate")
+      ```
+
+  * `def get_account_info()`
+
+    * `RuntimeError` should be replaced with `MissingParameterException`
+
+      ```python
+      try:
+          self.init_logic.fetch_account_from_server(self.conjurrc_data)
+      # pylint: disable=broad-except,logging-fstring-interpolation
+      except Exception as error:
+          # pylint: disable=line-too-long,logging-fstring-interpolation
+          logging.warning(f"Unable to fetch the account from the Conjur server. Reason: {error}")
+          # If there was a problem fetching the account from the server, we will request one
+          conjurrc_data.account = input("Enter the Conjur account name (required): ").strip()
+      
+          if conjurrc_data.account is None or conjurrc_data.account == '':
+              # pylint: disable=raise-missing-from
+              raise RuntimeError("Error: account is required")
+      ```
+
+  * `def __ensure_overwrite_file()`
+
+    * `Exception` should be replaced with `ConfirmationException`
+
+      ```python
+      force_overwrite = input(f"File {config_file} exists. " \
+                              f"Overwrite? (Default=yes): ").strip()
+      if force_overwrite != '' and force_overwrite.lower() != 'yes':
+          raise Exception(f"Not overwriting {config_file}")
+      ```
+
+* init_logic.py
+
+  * `def get_certificate()`
+
+    * `Exception` should be replaced with `FailedToRetrieveCertificateException` 
+
+      ```python
+      try:
+          fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
+          logging.debug("Successfully fetched certificate")
+      except Exception as error:
+          raise Exception(f"Unable to retrieve certificate from {hostname}:{port}. " \
+                          f"Reason: {str(error)}") from error
+      ```
+
+* login_controller.py
+
+  * `def get_username()`
+
+    * `RuntimeError` should be replaced with `MissingRequiredParameterException`
+
+      ```python
+      self.credential_data.login = input("Enter your login name to log into Conjur: ").strip()
+      if self.credential_data.login == '':
+          # pylint: disable=raise-missing-from
+          raise RuntimeError("Error: Login name is required")
+      ```
+
+  * `def get_api_key()`
+
+    * `RuntimeError` should be replaced with `AuthenticationException`
```

Comment:
we have AuthenticationException and AuthenticationFailedException. Let's pick one

Reply:
Picked AuthenticationFailedException